client/man/sxrev.1.in
.TH SXREV "1" "September 2014" "sxrev @VERSION@" "Skylable SX Manual"
.SH NAME
sxrev \- manage file revisions
.SH SYNOPSIS
.B sxrev
\fI\,<command> \/\fR[\fI\,OPTIONS\/\fR]...
.SH DESCRIPTION
The volumes in Skylable SX clusters can keep multiple revisions of files. For example, if a volume was created with an option to keep 5 revisions, every time a specific file gets modified the previous copy will be preserved and the latest 5 versions of the file will be available for download. The tools such as sxcp or sxls will always operate on the latest revision. sxrev can be used to list and manage all revisions of a particular file.
.SH COMMANDS
The available commands are listed below. To get more information about a specific command run \fB'man sxrev-COMMAND'\fR.
.TP
.B list
List available revisions of a file.
.TP
.B copy
Copy a specific revision of a file.
.TP
.B delete
Delete a specific revision of a file.
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit
.SH SEE ALSO
\fBsxls\fR(1), \fBsxcp\fR(1), \fBsxrm\fR(1), \fBsxcat\fR(1), \fBsxinit\fR(1)


NEWS
1.0
---

This is the first stable release of Skylable SX. The changes
from 0.9 include:

  * New user manual covering all important cluster and
    client operations (see doc/manual/manual.pdf or 
    http://www.skylable.com/products/sx/manual)
  * Support for replacing and rebuilding of broken nodes (see
    the manual for instructions on cluster healing)
  * New tool 'sxmv' for moving and renaming files (see 'man sxmv')
  * 'sxcp --ignore-errors' (ignore errors during recursive operations
    and report them at the end)
  * 'sxcp --exclude/--include' (skip or process files matching specific
    patterns)
  * 'sxvol modify' (change configuration of existing volumes)
  * 'sxadm --resize' (proportionally grow or shrink the entire cluster)
  * 'sxadm node --check' improvements (better detect local problems,
    eg. with faulty hard drives)
  * 'sxadm node --extract' (recover local storage in emergency situations)
  * Many improvements and bugfixes

0.9 (Final Beta)
----------------

This is a final beta of Skylable SX! Changes in this release include:

  * New garbage collector implementation, both more robust and scalable
  * Full support for file revisions (see 'man sxrev' and 'man sxvol' for
    details on how to manage and work with the revisions)
  * "undelete" filter (see 'sxvol filter --info undelete')
  * Usability improvements, including new man pages for the client tools
  * Many minor improvements and bugfixes

0.4 (Fourth Beta)
-----------------

This release contains important enhancements and bugfixes, including:

  * Automatic rebalance when nodes are added, modified or removed
    from the cluster
  * aes256 filter now uses the AES-256-CBC-HMAC-512 mode and bcrypt for
    key derivation
  * sxcp --one-file-system support (to not cross filesystem boundaries)
  * sxcp can now limit the number of connections (global and per host), see
    --full-help for usage information
  * Many usability enhancements in the client tools
  * Other improvements and bugfixes

0.3 (Third Beta)
----------------

The key features and improvements in this release include:

  * Support for using a custom port number (configure with sxsetup --advanced).
    This also allows sxsetup to run without root privileges when SX is
    configured to run on a port higher than 1024.
  * sxcp now displays a progress bar while transferring files
  * sxcp --bwlimit support (to set a bandwidth usage limit)
  * configure --with-system-libs support (to force the use of system libraries
    instead of those bundled in 3rdparty/)
  * Many minor improvements and bugfixes

0.2 (Second Beta)
-----------------

Changes in this release include:

  * Support for building with either NSS or OpenSSL
  * Minimum Curl version supported is now 7.34.0 (configure will use the
    embedded version if the system's one is not supported - this allows
    out-of-box build on CentOS 6)
  * aes256 filter now generates per-volume salts; in order to preserve
    deduplication support across different volumes using the same password
    you can force a static salt, eg. --filter aes256="salt:af6a812ad0b1cb22"
  * sxinit --force-reinit support (to recreate the entire local cluster config)
  * sxinit --list support (to list all configured clusters and profiles)
  * sxinit --alias support (to assign an alias for easy access), see
    https://wiki.skylable.com/wiki/Sxinit#Aliases
  * sxls -l now also lists filters used by volumes
  * Other bugfixes and improvements

0.1 (First Beta)
----------------

Initial public release of Skylable SX.


client/src/tools/rev/sxrev.c
/*
 *  Copyright (C) 2012-2014 Skylable Ltd. <info-copyright@skylable.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 2 as
 *  published by the Free Software Foundation.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 *  Special exception for linking this software with OpenSSL:
 *
 *  In addition, as a special exception, Skylable Ltd. gives permission to
 *  link the code of this program with the OpenSSL library and distribute
 *  linked combinations including the two. You must obey the GNU General
 *  Public License in all respects for all of the code used other than
 *  OpenSSL. You may extend this exception to your version of the program,
 *  but you are not obligated to do so. If you do not wish to do so, delete
 *  this exception statement from your version.
 */

#include "default.h"
#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <termios.h>

#include "sx.h"
#include "cmd_main.h"
#include "cmd_list.h"
#include "cmd_copy.h"
#include "cmd_delete.h"
#include "version.h"
#include "libsx/src/misc.h"
#include "bcrumbs.h"

static sxc_client_t *sx = NULL;

static int is_sx(const char *p) {
    return strncmp(p, "sx://", 5) == 0 || strncmp(p, SXC_ALIAS_PREFIX, strlen(SXC_ALIAS_PREFIX)) == 0;
}

static void sighandler(int signal)
{
    struct termios tcur;
    if(sx)
	sxc_shutdown(sx, signal);

    /* work around for ctrl+c during getpassword() in the aes filter */
    tcgetattr(0, &tcur);
    tcur.c_lflag |= ECHO;
    tcsetattr(0, TCSANOW, &tcur);

    fprintf(stderr, "Process interrupted\n");
    exit(1);
}

static sxc_file_t *make_sxfile(sxc_cluster_t **cluster, const char *host, const char *profile, const char *volume, const char *path, const char *rev) {
    sxc_file_t *file;

    if(!volume || !path) {
	fprintf(stderr, "ERROR: Bad file path\n");
	return NULL;
    }

    if(!*cluster)
	*cluster = sxc_cluster_load_and_update(sx, host, profile);

    if(!*cluster) {
	fprintf(stderr, "ERROR: Failed to load config for %s: %s\n", host, sxc_geterrmsg(sx));
	if(strstr(sxc_geterrmsg(sx), SXBC_TOOLS_CFG_ERR))
	    fprintf(stderr, SXBC_TOOLS_CFG_MSG, host, host);
	else if(strstr(sxc_geterrmsg(sx), SXBC_TOOLS_CONN_ERR))
	    fprintf(stderr, SXBC_TOOLS_CONN_MSG);
	return NULL;
    }

    file = sxc_file_remote(*cluster, volume, path, rev);
    if(!file)
	fprintf(stderr, "ERROR: Failed to create file object: %s\n", sxc_geterrmsg(sx));

    return file;
}


int main(int argc, char **argv) {
    sxc_file_t *file = NULL, *destfile = NULL;
    char *config_dir = NULL, *filter_dir = NULL, *selected_rev = NULL;
    const char *src;
    sxc_logger_t log;
    sxc_cluster_t *cluster = NULL, *destcluster = NULL;
    sxc_revlist_t *revs = NULL;
    sxc_uri_t *uri = NULL;
    unsigned int i, match = 0, debug = 0;
    enum { OPNONE, OPMAIN, OPLIST, OPCOPY, OPDELETE } op = OPNONE;
    struct main_args_info main_args;
    struct list_args_info list_args;
    struct copy_args_info copy_args;
    struct delete_args_info delete_args;
    int ret = 1;

    if(argc < 2) {
	main_cmdline_parser_print_help();
	fprintf(stderr, "ERROR: No command specified\n");
	return 1;
    }

    if(!strcmp(argv[1], "list")) {
	if(list_cmdline_parser(argc - 1, &argv[1], &list_args)) {
	    printf("\n");
	    list_cmdline_parser_print_help();
	    printf("\n");
	    fprintf(stderr, "ERROR: Invalid syntax or usage\n");
	    goto err;
	}

	op = OPLIST;

	if(list_args.version_given) {
	    printf("%s %s\n", MAIN_CMDLINE_PARSER_PACKAGE, SRC_VERSION);
	    ret = 0;
	    goto err;
	}

	if(list_args.inputs_num != 1) {
	    list_cmdline_parser_print_help();
	    printf("\n");
	    fprintf(stderr, "ERROR: Invalid number of arguments\n");
	    goto err;
	}

	debug = list_args.debug_flag;
	if(list_args.filter_dir_given)
	    filter_dir = list_args.filter_dir_arg;
	if(list_args.config_dir_given)
	    config_dir = list_args.config_dir_arg;
	src = list_args.inputs[0];
	if(!is_sx(src)) {
	    list_cmdline_parser_print_help();
	    printf("\n");
	    fprintf(stderr, "ERROR: Bad SX file\n");
	}
    } else if(!strcmp(argv[1], "copy")) {
	if(copy_cmdline_parser(argc - 1, &argv[1], &copy_args)) {
	    printf("\n");
	    copy_cmdline_parser_print_help();
	    printf("\n");
	    fprintf(stderr, "ERROR: Invalid syntax or usage\n");
	    goto err;
	}

	op = OPCOPY;

	if(copy_args.version_given) {
	    printf("%s %s\n", MAIN_CMDLINE_PARSER_PACKAGE, SRC_VERSION);
	    ret = 0;
	    goto err;
	}

	if(copy_args.inputs_num != 2) {
	    copy_cmdline_parser_print_help();
	    printf("\n");
	    fprintf(stderr, "ERROR: Invalid number of arguments\n");
	    goto err;
	}

	debug = copy_args.debug_flag;
	if(copy_args.filter_dir_given)
	    filter_dir = copy_args.filter_dir_arg;
	if(copy_args.config_dir_given)
	    config_dir = copy_args.config_dir_arg;
	src = copy_args.inputs[0];
	if(!is_sx(src)) {
	    copy_cmdline_parser_print_help();
	    printf("\n");
	    fprintf(stderr, "ERROR: Bad SX file\n");
	}
	if(copy_args.rev_given)
	    selected_rev = copy_args.rev_arg;
    } else if(!strcmp(argv[1], "delete")) {
	if(delete_cmdline_parser(argc - 1, &argv[1], &delete_args)) {
	    printf("\n");
	    delete_cmdline_parser_print_help();
	    printf("\n");
	    fprintf(stderr, "ERROR: Invalid syntax or usage\n");
	    goto err;
	}

	op = OPDELETE;

	if(delete_args.version_given) {
	    printf("%s %s\n", MAIN_CMDLINE_PARSER_PACKAGE, SRC_VERSION);
	    ret = 0;
	    goto err;
	}

	if(delete_args.inputs_num != 1) {
	    delete_cmdline_parser_print_help();
	    printf("\n");
	    fprintf(stderr, "ERROR: Invalid number of arguments\n");
	    goto err;
	}

	debug = delete_args.debug_flag;
	if(delete_args.filter_dir_given)
	    filter_dir = delete_args.filter_dir_arg;
	if(delete_args.config_dir_given)
	    config_dir = delete_args.config_dir_arg;
	src = delete_args.inputs[0];
	if(!is_sx(src)) {
	    delete_cmdline_parser_print_help();
	    printf("\n");
	    fprintf(stderr, "ERROR: Bad SX file\n");
	}
	if(delete_args.rev_given)
	    selected_rev = delete_args.rev_arg;
    } else {
	if(argc > 2) {
	    fprintf(stderr, "ERROR: Unknown command '%s'\n", argv[1]);
	    goto err;
	}

	if(main_cmdline_parser(argc, argv, &main_args))
	    goto err;

	if(main_args.version_given) {
	    printf("%s %s\n", MAIN_CMDLINE_PARSER_PACKAGE, SRC_VERSION);
	    op = OPMAIN;
	    ret = 0;
	} else
	    fprintf(stderr, "ERROR: Unknown command '%s'\n", argv[1]);
	goto err;
    }

    signal(SIGINT, sighandler);
    signal(SIGTERM, sighandler);

    if(!(sx = sxc_init(SRC_VERSION, sxc_default_logger(&log, argv[0]), sxc_input_fn, NULL)))
	goto err;

    sxc_set_debug(sx, debug);
    if(config_dir && sxc_set_confdir(sx, config_dir)) {
        fprintf(stderr, "ERROR: Could not set configuration directory %s: %s\n", config_dir, sxc_geterrmsg(sx));
	goto err;
    }

    /* At this time there is no use for filters here but they are loaded anyway for consistency */
    if(filter_dir) {
	filter_dir = strdup(filter_dir);
    } else {
	const char *pt = sxi_getenv("SX_FILTER_DIR");
	if(pt)
	    filter_dir = strdup(pt);
	else
	    filter_dir = strdup(SX_FILTER_DIR);
    }
    if(!filter_dir) {
	fprintf(stderr, "ERROR: Failed to set filter dir\n");
	goto err;
    }
    sxc_filter_loadall(sx, filter_dir);
    free(filter_dir);

    if(op == OPCOPY) {
	const char *dst = copy_args.inputs[1];
	if(is_sx(dst)) {
	    sxc_uri_t *desturi = sxc_parse_uri(sx, dst);
	    if(!desturi) {
		fprintf(stderr, "ERROR: Bad uri %s: %s\n", dst, sxc_geterrmsg(sx));
		goto err;
	    }
	    destfile = make_sxfile(&destcluster, desturi->host, desturi->profile, desturi->volume, desturi->path, NULL);
	    sxc_free_uri(desturi);
	    if(!destfile)
		goto err;
	} else {
	    destfile = sxc_file_local(sx, dst);
	    if(!destfile) {
		fprintf(stderr, "ERROR: Failed to create destination file object: %s\n", sxc_geterrmsg(sx));
		goto err;
	    }
	}
    }

    if(!(uri = sxc_parse_uri(sx, src))) {
	fprintf(stderr, "ERROR: Bad uri %s: %s\n", src, sxc_geterrmsg(sx));
	goto err;
    }

    if(!(file = make_sxfile(&cluster, uri->host, uri->profile, uri->volume, uri->path, NULL)))
	goto err;
    revs = sxc_revisions(file);
    sxc_file_free(file);

    if(!revs) {
	fprintf(stderr, "ERROR: Failed to retrieve file revisions: %s\n", sxc_geterrmsg(sx));
	goto err;
    }

    if(!selected_rev)
	printf("Revisions for file %s (most recent first):\n", src);
    for(i=0; i<revs->count; i++) {
	const sxc_revision_t *rev = revs->revisions[i];
	struct tm *gt = gmtime(&rev->created_at);
	if(selected_rev) {
	    if(!strcmp(rev->revision, selected_rev)) {
		match = 1;
		break;
	    }
	} else
	    printf("%u.\t%04d-%02d-%02d %02d:%02d size:%llu rev:\"%s\"\n",
		   i+1,
		   gt->tm_year + 1900,
		   gt->tm_mon + 1,
		   gt->tm_mday,
		   gt->tm_hour,
		   gt->tm_min,
		   (long long)rev->file_size,
		   rev->revision);
    }

    if(op == OPLIST) {
	ret = 0;
	goto err;
    }

    if(!selected_rev) {
	char choice[32];
	printf("Choose revision to %s: ", op == OPCOPY ? "copy" : "delete");
	if(fgets(choice, sizeof(choice), stdin)) {
	    i = atoi(choice) - 1;
	    if(i<revs->count) {
		selected_rev = revs->revisions[i]->revision;
		match = 1;
	    }
	}
    }

    if(!match) {
	fprintf(stderr, "The specified revision does not exist\n");
	ret = 2;
	goto err;
    }


    file = make_sxfile(&cluster, uri->host, uri->profile, uri->volume, uri->path, selected_rev);
    if(!file)
	goto err;

    if(op == OPCOPY && !sxc_copy_sxfile(file, destfile, 0))
	ret = 0;

    if(op == OPDELETE && !sxc_remove_sxfile(file))
	ret = 0;

    if(ret)
	fprintf(stderr, "%s operation failed: %s\n", op == OPCOPY ? "Copy" : "Delete", sxc_geterrmsg(sx));
    else 
	fprintf(stderr, "%s operation completed successfully\n", op == OPCOPY ? "Copy" : "Delete");

    sxc_file_free(file);

 err:
    signal(SIGINT, SIG_IGN);
    signal(SIGTERM, SIG_IGN);

    sxc_file_free(destfile);
    sxc_revisions_free(revs);
    sxc_free_uri(uri);
    sxc_cluster_free(destcluster);
    sxc_cluster_free(cluster);
    sxc_shutdown(sx, 0);
    if(op == OPMAIN)
	main_cmdline_parser_free(&main_args);
    else if(op == OPLIST)
	list_cmdline_parser_free(&list_args);
    else if(op == OPCOPY)
	copy_cmdline_parser_free(&copy_args);
    else if(op == OPDELETE)
	delete_cmdline_parser_free(&delete_args);

    return ret;
}
